java-frontend/src/main/java/org/sonar/java/se/checks/InvariantReturnCheck.java

```diff
@@ -0,0 +1,159 @@
+    private final Multimap<Class<? extends Constraint>, Constraint> methodConstraints = ArrayListMultimap.create();
+    private final List<ReturnStatementTree> returnStatementTrees;
+    private int endPaths = 0;
+    private boolean methodToCheck = false;
+    private boolean returnImmutableType = false;
+    private boolean avoidRaisingConstraintIssue = false;
+
+    public MethodInvariantContext(MethodTree methodTree) {
+      this.methodTree = methodTree;
+      TypeTree returnType = methodTree.returnType();
+      this.returnStatementTrees = extractReturnStatements(methodTree);
+      methodToCheck = !isConstructorOrVoid(returnType) && returnStatementTrees.size() > 1;
+      returnImmutableType = methodToCheck && (returnType.symbolType().isPrimitive() || returnType.symbolType().is("java.lang.String"));
+    }
+
+    private static boolean isConstructorOrVoid(@Nullable TypeTree returnType) {
+      return returnType == null || returnType.symbolType().isVoid();
+    }
+
+    private static List<ReturnStatementTree> extractReturnStatements(MethodTree methodTree) {
+      ReturnExtractor visitor = new ReturnExtractor();
+      methodTree.accept(visitor);
+      return visitor.returns;
+    }
+  }
+
+  private Deque<MethodInvariantContext> methodInvariantContexts = new LinkedList<>();
+
+  @Override
+  public void init(MethodTree methodTree, CFG cfg) {
+    methodInvariantContexts.push(new MethodInvariantContext(methodTree));
+  }
+
+  @Override
+  public void checkEndOfExecutionPath(CheckerContext context, ConstraintManager constraintManager) {
+    MethodInvariantContext methodInvariantContext = methodInvariantContexts.peek();
+    if (!methodInvariantContext.methodToCheck) {
+      return;
+    }
+    SymbolicValue exitValue = context.getState().exitValue();
+
+    if (exitValue != null) {
+      methodInvariantContext.endPaths++;
+      methodInvariantContext.symbolicValues.add(exitValue);
+      PMap<Class<? extends Constraint>, Constraint> constraints = context.getState().getConstraints(exitValue);
+      if (constraints != null) {
+        constraints.forEach(methodInvariantContext.methodConstraints::put);
+      } else {
+        // Relational SV or NOT SV : we can't say anything.
+        methodInvariantContext.avoidRaisingConstraintIssue = true;
+      }
+    }
+  }
+
+  @Override
+  public void checkEndOfExecution(CheckerContext context) {
+    reportIssues();
+  }
+
+  @Override
+  public void interruptedExecution(CheckerContext context) {
+    reportIssues();
+  }
+
+  private void reportIssues() {
+    MethodInvariantContext methodInvariantContext = methodInvariantContexts.pop();
+    if (!methodInvariantContext.methodToCheck) {
+      return;
+    }
+    if (methodInvariantContext.returnImmutableType && methodInvariantContext.symbolicValues.size() == 1 && methodInvariantContext.endPaths > 1) {
+      report(methodInvariantContext);
+    }
+    if(!methodInvariantContext.avoidRaisingConstraintIssue) {
```

Comment:
ok case added. but not triggerable because of filtering done in SE issue reporting.